akka-http/src/main/scala/akka/http/scaladsl/server/directives/FileUploadDirectives.scala

```diff
@@ -32,28 +35,76 @@ trait FileUploadDirectives {
    *
    * @group fileupload
    */
+  @deprecated("Deprecated in favor of storeUploadedFile which allows to specify a file to store the upload in.", "10.0.11")
   def uploadedFile(fieldName: String): Directive1[(FileInfo, File)] =
+    storeUploadedFile(fieldName, _ ⇒ File.createTempFile("akka-http-upload", ".tmp")).tmap(Tuple1(_))
+
+  /**
+   * Streams the bytes of the file submitted using multipart with the given file name into a designated file on disk.
+   * If there is an error writing to disk the request will be failed with the thrown exception, if there is no such
+   * field the request will be rejected, if there are multiple file parts with the same name, the first one will be
+   * used and the subsequent ones ignored.
+   *
+   * @group fileupload
+   */
+  @ApiMayChange
+  def storeUploadedFile(fieldName: String, destFn: FileInfo ⇒ File): Directive[(FileInfo, File)] =
     extractRequestContext.flatMap { ctx ⇒
       import ctx.executionContext
       import ctx.materializer
 
       fileUpload(fieldName).flatMap {
         case (fileInfo, bytes) ⇒
 
-          val destination = File.createTempFile("akka-http-upload", ".tmp")
-          val uploadedF: Future[(FileInfo, File)] = bytes.runWith(FileIO.toPath(destination.toPath))
-            .map(_ ⇒ (fileInfo, destination))
-
-          onComplete[(FileInfo, File)](uploadedF).flatMap {
+          val dest = destFn(fileInfo)
+          val uploadedF: Future[(FileInfo, File)] =
+            bytes
+              .runWith(FileIO.toPath(dest.toPath))
+              .map(_ ⇒ (fileInfo, dest))
+              .recoverWith {
+                case ex ⇒
+                  dest.delete()
+                  throw ex
+              }
+
+          onSuccess(uploadedF)
+      }
+    }
 
-            case Success(uploaded) ⇒
-              provide(uploaded)
+  /**
+   * Streams the bytes of the file submitted using multipart with the given field name into designated files on disk.
+   * If there is an error writing to disk the request will be failed with the thrown exception, if there is no such
+   * field the request will be rejected. Stored files are cleaned up on exit but not on failure.
```

Comment:
Does storeUploadedFiles Directive reject request if field is empty? 
It seems there is no implementation of it